### ISSUE TITLE
Dropping PHP 5.5 in SuiteCRM 7.10

### ISSUE THREAD

**Issue by connorshea** [salesagility/SuiteCRM#7463]
#### Issue
There are a number of things that are causing problems with continuing to support PHP 5.5 on SuiteCRM's 7.10 branch. I'd like to have a discussion about whether dropping PHP 5.5 in a patch release is viable.

There are a few reasons I think we should do this:
- 7.11 dropped 5.5 support, which makes some backports to 7.10 more difficult.
- 7.10.x is an LTS and will still be supported for another full year from now, which means the backporting issues will only get worse as time goes on.
- Many libraries have dropped support for 5.5, most notably Codeception in (if I remember correctly) v2.4.2. This means we're missing [a lot of new features and fixes](https://codeception.com/changelog), and the 7.11 branch is on v3 while 7.10 isn't even on the latest 2.x release.
- php-cs-fixer also does not support PHP 5.5, which makes it impossible to run CI linting on the hotfix-7.10 branch. This means if we enable linting on hotfix, when 7.10 changes are merged into hotfix every month there will likely be linting failures introduced that will need to be fixed during the merge process. This makes both linting and merging significantly more painful.
- PHP 5.5 is unsupported by the PHP upstream project and has been EOL since **June 2016**, 3 years! https://www.php.net/supported-versions.php
- 5.5 is another version we have to test in CI, which makes CI take longer.
- PHP 5.5 and 5.6 should mostly be compatible with each other(?) so upgrading shouldn't be a significant problem for most users.

The downsides:
- Dropping support for a PHP version in a patch release in not really something that should normally be done, but as long as it's well-communicated to users in the changelog it will hopefully not cause any significant surprises for users upgrading their CRM.
- If anyone is still using PHP 5.5, this will force them to either stay on the release before the one where we drop support or upgrade their PHP version. This is less than ideal, but PHP 5.5 hasn't had security updates in over 3 years. They should have already upgraded a long time ago.

PHP 5.5 has been a significant thorn in my side as I've been trying to update the test suite and improve the project's code quality, specifically because of Codeception and php-cs-fixer dropping support for it. I imagine there are other libraries that have also dropped support that I'm not currently aware of (I know PHPUnit did, but it seems like they drop support pretty aggressively anyway).

I know doing this would violate an implicit (or explicit? I don't know if it's formalized anywhere) promise with users that patch releases wouldn't involve dropping support for PHP versions, but I think it's important for the continued development of the project over the next year, otherwise we'll be spending a lot of time working around a version of PHP that - at least I hope - very few users are actually running.

**Comment by pgorod** [comment]
I agree with the proposal, but if for LTS reasons we don't drop support in the product yet, maybe we can just drop it from the tests. 

So we could maybe instruct people with really old PHP's to run the 5.5 tests themselves, if they care enough about it (but if they cared, they would have upgraded PHP long ago, right?).

**Comment by connorshea** [comment]
The problem with that is that the supported PHP versions are specified in `composer.json`. We can't upgrade codeception/php-cs-fixer without dropping support for PHP 5.5 from the `composer.json`, and then I don't think 7.10 would be able to be run on 5.5.

**Comment by mcarpenterjr** [comment]
I think this one of those issues where "you do" or "you don't" and trying to support both sides of the fence most likely will cause confusion and frustration. Overall security should be a priority, but there are probably plenty of user cases out there where suite's not running in a container or an isolated environment and making the move to php7 for some might cause conflict in other services/applications. Not that should ultimately be suite's concern, just something to be prepared to make a case for.

Does suite know the depth of impact on the code base form 5 to 7?

**Comment by connorshea** [comment]
To be clear, this’d only drop 5.5. 5.6 would still be supported, so upgrading to 7 wouldn’t be strictly necessary.

**Comment by mcarpenterjr** [comment]
Oh, in that case, I would suspect end-user impact should be minimal, the difficult jump at that point was from 5.4 to 5.5, some dependency versions for 5.5 aren't supported on Ubuntu 12 and older. AFAIK there shouldn't be much conflict. Thanks for clearing that up.

**Comment by Dillon-Brown** [comment]
Interesting usage statistics for PHP versions: https://w3techs.com/technologies/details/pl-php/all/all. I definitely agree that dropping 5.5 support would make development easier when dealing with the hotfix-7.10.x and hotfix branches. Although as Pedro mentioned, we typically don't drop support in minor releases.

**Comment by connorshea** [comment]
@Dillon-Brown I think it's worth noting that Wordpress causes these stats to lean pretty heavily toward older releases, since it's easy to set up and a lot of people don't upgrade the underlying PHP on their server. I don't think that should stop us from dropping 5.5. (plus it's not specific on what exact version of PHP 5 people are on, it could be 5.2, 5.5, or 5.6)

**Comment by samus-aran** [comment]
Hey @connorshea we can understand the need to raise this and I agree that it certainly would make testing, processes etc a bit easier. But we have to be careful not to pigeon hole our community to only those active in the GitHub repo. We have hundreds of thousands of downloads all in different environments all with their own limitations (be it business, or technological). So if we were to even consider this upgrade then we need to, as you said, communicate but first and foremost... actually ask everyone (as much as possible) what do they currently use. And that is a time consuming task which, still would only give us a glimpse into our community's environments. 

I'm not saying we shouldn't but one we would need to set a more realistic deadline for an answer for this and not rush the decision. Not only would we be going against 'LTS' but we are disrupting people's workflow by throwing this possible spanner in the works - and no, giving them "oh, we are "encouraging" you to upgrade your entire system" Is the best approach to go with haha.

So, at the moment it's a No unless there is enough evidence from our community that doing this will not have too much of an impact on them. Yes, I think we should look at gathering this evidence anyway to help our own processes. I'm open to suggestions on how to go about doing that send us a message community@suitecrm.com

**Comment by connorshea** [comment]
@samus-aran 

A few ideas for getting the word out:
- Add a notice to the 7.10.19 and 7.10.20 release notes that 7.10.21 will drop PHP 5.5 and see if anyone complains, this should give people at least a month if not two months of warning
- Pin this issue?
- Post on the forums about dropping PHP 5.5
- Potentially, we could add a warning for anyone on PHP 5.5 that'll show up when they install 7.10.20 that'd say PHP 5.5 will be dropped in 7.10.21?

This one may be excessive, but maybe even put a banner on suitecrm.com about dropping PHP 5.5?

On all of these, we'd include links to the GitHub issue, an email to contact if you absolutely need PHP 5.5 support, and a link to the forum thread on the topic.

Based on the packagist graphs, it doesn't look PHP 5.5 has particularly significant usage numbers, thankfully. I think if you're still on PHP 5.5, it's unlikely you're upgrading the CRM much anyway.

**Comment by connorshea** [comment]
@samus-aran *poke* we should probably start on this now if we want to drop it for 7.10.21 :)

**Comment by connorshea** [comment]
I still need to post about this on the forums, if anyone wants to do that for me I'd be grateful :) Otherwise I'll try to get it done this week, please pester me if I don't 🙇

**Comment by connorshea** [comment]
Here's the forum post: https://suitecrm.com/suitecrm/forum/suitecrm-7-0-discussion/27442-dropping-support-for-php-5-5-in-the-next-7-10-x-release

**Comment by connorshea** [comment]
@samus-aran @Dillon-Brown seeing as we aren't running the tests on PHP 5.5 anymore and the forum post hasn't gotten any responses, should we drop 5.5 in the upcoming release or still wait until the one after?

**Comment by connorshea** [comment]
@Dillon-Brown @samus-aran obligatory ping for removing PHP 5.5 support in the next release. :)

**Comment by maxxatgit** [comment]
Greetings all, I rarely post, but here's my 2 cents:
Cpanel makes it quite easy to run several different versions of php on the same server, and I think DirectAdmin does too.
Sometimes it's a burden on users to have to update php (in this case, they're not using a control panel, and they are running other apps that *require* php5.5) but it's always a burden on the developers to make new software work on older php versions. I was extremely happy when Suite first became compatible with php 7.
I'm all for dropping php5.5 support, the php devs already have.

**Comment by tsmgeek** [comment]
Same here, drop it plesase.

**Comment by chris001** [comment]
The app should support PHP versions which are Actively Supported or Security Supported, as of right now 7.2, 7.3, and 7.4.
Drop support for PHP versions which are EOL/dead, 5.5, 5.6, 7.0, 7.1:
https://www.php.net/supported-versions.php
